presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/README.md

```diff
@@ -0,0 +1,148 @@
+actual plan. The simplest possible example of this would be asserting that a
+`SELECT name FROM people` query has an `OutputNode` with an output `Symbol`
+that actually represents the data in the `name` column of the `people` table.
+
+In order for the plan testing framework to be independent of how Presto
+allocates `Symbols`, `Symbols` are given an alias where they originate, and
+must be referenced by that alias further up the plan. The value being assigned
+to a `Symbol` can either be matched against something that is independent of a
+`Symbol` (e.g., a `ColumnHandle` in a `TableScanNode`) or a value that is
+dependent on a previously aliased symbol (e.g., an `Expression` in a
+`ProjectNode`). Matching against the right side of the assignment ensures that
+the plan matching framework is independent of the implementation of the
+`SymbolAllocator`.
+
+## Two types of tests
+
+### Testing planner code
+
+Planner code should be tested by comparing the unoptimized plan to a set of
+expectations. For various reasons related to how `Symbols` are handled in
+unoptimized plans, this is frequently impossible. The minimum set of optimizers
+that needs to be applied for the plan matching framework to function is likely
+to be some subset of the following:
+
+- `UnaliasSymbolReferences`
+- `PruneIdentityProjections`
+- `PruneUnreferencedOutputs`
+
+`TestQuantifiedComparison` illustrates unit testing planner code.
+
+### Testing optimizer code
+
+Optimizer code will generally need to be subjected to two types of tests:
+
+1. Unit tests on the minimum set of optimizers needed for the optimizer to
+   function.
+2. A test using the full set of optimizers that ensures that the optimizer
+   continues to function after the introduction of new optimizers or reordering
+   of existing optimizers.
+
+`TestMergeWindows` illustrates both types of tests and goes further into the
+rationale for both types of tests.
+
+## Internals
+
+### The PlanMatchingVisitor
+
+The `PlanMatchingVisitor` is responsible for visting the actual plan and
+ensuring that the actual plan conforms to the expectations expressed in the
+test.
+
+### PlanMatchPatterns
+
+`PlanMatchPatterns` contain a list of source `PlanMatchPatterns` and a list of
+`Matchers`. The `PlanMatchingVisitor` is responsible for traversing the actual
+plan and making sure that the structure mirrors that of the tree of
+`PlanMatchPattens`, and that the `Matchers` in the `PlanMatchPatterns` match
+the nodes.
+
+Conceptually, this is relatively simple, but the need for flexibility adds a
+wrinkle. A `PlanMatchPattern` contains a field `anyTree`, which signifies that
+it can be matched against any `PlanNode` further down the tree.
+
+### Matchers
+
+The `Matcher` interface is used to express expectations about specific nodes.
+`Matchers` are applied in two steps when traversing the plans: As the matching
+algorithm passes downwards from the root of the tree it applies the
```

Comment:
...are applied as follows when traversing the plans:
Bullet for downwards
Bullet for upwards